### PR TITLE
make aad obo autoconfiguration depend on aad.client-id

### PR DESCRIPTION
**Pull request by saragluna**
In the current implementation, when users add the aad starter as a dependency while not configuring any aad properties, the obo autoconfiguration will complain that aad-clientid must be set. 
This PR is to fix this issue.

### CODE REVIEW

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAuthenticationProperties.java [hunk 1]
```diff
@@ -20,14 +20,17 @@
...
 import java.util.concurrent.TimeUnit;
 
+import static com.azure.spring.autoconfigure.aad.AADAuthenticationProperties.AAD_PREFIX;
+
 /**
  * Configuration properties for Azure Active Directory Authentication.
  */
 @Validated
-@ConfigurationProperties("azure.activedirectory")
+@ConfigurationProperties(AAD_PREFIX)
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAuthenticationProperties.java:29]: Can we keep the same style? All use `AAD_PREFIX`, or not use it at all.

Ofcourse, we can do it after release.

File: sdk/spring/azure-spring-boot/src/test/java/com/azure/spring/aad/webapi/AADResourceServerOboConfigurationTest.java [hunk 2]
```diff
@@ -24,6 +26,17 @@
             "azure.activedirectory.client-id=fake-client-id",
             "azure.activedirectory.client-secret=fake-client-secret");
 
+    @Test
+    public void testWithoutAnyPropertiesSet() {
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/test/java/com/azure/spring/aad/webapi/AADResourceServerOboConfigurationTest.java:30]: Do we need test like `testWithoutRequiredPropertiesSet`?

**Review by chenrujun — CHANGES_REQUESTED** [review]

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAuthenticationProperties.java (hunk 1 above)
Comment (saragluna) [reply sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAuthenticationProperties.java:29]: Which one do you prefer?

File: sdk/spring/azure-spring-boot/src/test/java/com/azure/spring/aad/webapi/AADResourceServerOboConfigurationTest.java (hunk 2 above)
Comment (saragluna) [reply sdk/spring/azure-spring-boot/src/test/java/com/azure/spring/aad/webapi/AADResourceServerOboConfigurationTest.java:30]: Good point. I will add that.

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAuthenticationProperties.java (hunk 1 above)
Comment (chenrujun) [reply sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAuthenticationProperties.java:29]: I prefer not use it at all. If we want to get some property, we can search by `CTRL + SHIFT + F` to get them all.

**Review by chenrujun — APPROVED** [review]
LGTM.

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAuthenticationProperties.java (hunk 1 above)
Comment (saragluna) [reply sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AADAuthenticationProperties.java]: okay

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapi/AADResourceServerOboConfiguration.java [hunk 3]
```diff
@@ -33,6 +35,8 @@
 @ConditionalOnResource(resources = "classpath:aad.enable.config")
 @EnableConfigurationProperties({ AADAuthenticationProperties.class })
 @ConditionalOnClass({ BearerTokenAuthenticationToken.class, OAuth2LoginAuthenticationFilter.class })
+@ConditionalOnProperty(prefix = "azure.activedirectory", value = "client-id")
+@ConditionalOnExpression("!'${azure.activedirectory.authorization-clients}'.isEmpty()")
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapi/AADResourceServerOboConfiguration.java:39]: Good point. 👍

File: sdk/spring/azure-spring-boot/src/test/java/com/azure/spring/aad/webapi/AADResourceServerOboConfigurationTest.java [hunk 4]
```diff
@@ -24,6 +26,30 @@
...
+        new WebApplicationContextRunner()
+            .withUserConfiguration(AADResourceServerOboConfiguration.class)
+            .run(context -> {
+                assertThat(context).doesNotHaveBean(AADAuthenticationProperties.class);
+                assertThat(context).doesNotHaveBean(ClientRegistrationRepository.class);
+                assertThat(context).doesNotHaveBean(OAuth2AuthorizedClientRepository.class);
+            });
+    }
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/test/java/com/azure/spring/aad/webapi/AADResourceServerOboConfigurationTest.java:38]: We'd better add more tests: 
1. `client-id` configured, but no `authorization-clients`.
2. `client-id` not configured, but `authorization-clients` configured.

**Review by chenrujun — CHANGES_REQUESTED** [review]

File: sdk/spring/azure-spring-boot/src/test/java/com/azure/spring/aad/webapi/AADResourceServerOboConfigurationTest.java (hunk 4 above)
Comment (saragluna) [reply sdk/spring/azure-spring-boot/src/test/java/com/azure/spring/aad/webapi/AADResourceServerOboConfigurationTest.java:38]: Yes, I'm just about to add them.

Comment (chenrujun) [reply sdk/spring/azure-spring-boot/src/test/java/com/azure/spring/aad/webapi/AADResourceServerOboConfigurationTest.java:38]: Current tests can not make sure `@ConditionalOnExpression("!'${azure.activedirectory.authorization-clients}'.isEmpty()")` take effect.

File: sdk/spring/azure-spring-boot/src/test/java/com/azure/spring/aad/webapi/AADResourceServerOboConfigurationTest.java [hunk 5]
```diff
@@ -24,24 +26,43 @@
...
+
     @Test
     public void testNotExistBearerTokenAuthenticationToken() {
         this.contextRunner
             .withUserConfiguration(AADResourceServerOboConfiguration.class)
             .withClassLoader(new FilteredClassLoader(BearerTokenAuthenticationToken.class))
-            .run(context -> {
-                assertThat(context).doesNotHaveBean("AADOAuth2OboAuthorizedClientRepository");
-            });
+            .run(context -> assertThat(context).doesNotHaveBean("AADOAuth2OboAuthorizedClientRepository"));
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/test/java/com/azure/spring/aad/webapi/AADResourceServerOboConfigurationTest.java:57]: `doesNotHaveBean` should use bean name.  `AADOAuth2OboAuthorizedClientRepository` is not a bean name.

My suggestion is to use 
```
doesNotHaveBean(Class<?> type, ApplicationContextAssert.Scope scope) {
```
instead.

File: sdk/spring/azure-spring-boot/src/test/java/com/azure/spring/aad/webapi/AADResourceServerOboConfigurationTest.java [hunk 6]
```diff
@@ -24,24 +26,43 @@
...
 
     @Test
     public void testNotExistOAuth2LoginAuthenticationFilter() {
         this.contextRunner
             .withUserConfiguration(AADResourceServerOboConfiguration.class)
             .withClassLoader(new FilteredClassLoader(OAuth2LoginAuthenticationFilter.class))
-            .run(context -> {
-                assertThat(context).doesNotHaveBean("AADOAuth2OboAuthorizedClientRepository");
-            });
+            .run(context -> assertThat(context).doesNotHaveBean("AADOAuth2OboAuthorizedClientRepository"));
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/test/java/com/azure/spring/aad/webapi/AADResourceServerOboConfigurationTest.java:65]: Same here.

File: sdk/spring/azure-spring-boot/src/test/java/com/azure/spring/aad/webapi/AADResourceServerOboConfigurationTest.java [hunk 7]
```diff
@@ -24,24 +26,43 @@
...
+
+    @Test
+    public void testWithRequiredPropertiesSet() {
+        new WebApplicationContextRunner()
+            .withUserConfiguration(AADResourceServerOboConfiguration.class)
+            .withPropertyValues("azure.activedirectory.client-id=fake-client-id")
+            .run(context -> {
+                assertThat(context).hasSingleBean(AADAuthenticationProperties.class);
+                assertThat(context).hasSingleBean(ClientRegistrationRepository.class);
+                assertThat(context).hasSingleBean(OAuth2AuthorizedClientRepository.class);
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot/src/test/java/com/azure/spring/aad/webapi/AADResourceServerOboConfigurationTest.java:48]: To keep same style, use
```
hasSingleBean(Class<?> type, ApplicationContextAssert.Scope scope)
```

**Review by chenrujun — CHANGES_REQUESTED** [review]

**Review by chenrujun — APPROVED** [review]
LGTM.

**Pull request merged**